pageserver/src/pgdatadir_mapping.rs

```diff
@@ -1331,56 +1363,58 @@ impl Timeline {
 
     /// Get cached size of relation if it not updated after specified LSN
     pub fn get_cached_rel_size(&self, tag: &RelTag, lsn: Lsn) -> Option<BlockNumber> {
-        let rel_size_cache = self.rel_size_cache.read().unwrap();
-        if let Some((cached_lsn, nblocks)) = rel_size_cache.map.get(tag) {
+        let rel_size_cache = self.rel_size_primary_cache.read().unwrap();
+        if let Some((cached_lsn, nblocks)) = rel_size_cache.get(tag) {
             if lsn >= *cached_lsn {
                 RELSIZE_CACHE_HITS.inc();
                 return Some(*nblocks);
             }
             RELSIZE_CACHE_MISSES_OLD.inc();
         }
+        let mut rel_size_cache = self.rel_size_replica_cache.lock().unwrap();
+        if let Some(nblock) = rel_size_cache.get(&(lsn, *tag)) {
+            RELSIZE_CACHE_HITS.inc();
+            return Some(*nblock);
+        }
         RELSIZE_CACHE_MISSES.inc();
         None
     }
 
     /// Update cached relation size if there is no more recent update
-    pub fn update_cached_rel_size(&self, tag: RelTag, lsn: Lsn, nblocks: BlockNumber) {
-        let mut rel_size_cache = self.rel_size_cache.write().unwrap();
-
-        if lsn < rel_size_cache.complete_as_of {
-            // Do not cache old values. It's safe to cache the size on read, as long as
-            // the read was at an LSN since we started the WAL ingestion. Reasoning: we
-            // never evict values from the cache, so if the relation size changed after
-            // 'lsn', the new value is already in the cache.
-            return;
-        }
-
-        match rel_size_cache.map.entry(tag) {
-            hash_map::Entry::Occupied(mut entry) => {
-                let cached_lsn = entry.get_mut();
-                if lsn >= cached_lsn.0 {
-                    *cached_lsn = (lsn, nblocks);
+    pub fn update_cached_rel_size(&self, tag: RelTag, version: Version<'_>, nblocks: BlockNumber) {
```

Comment:
Can we use a pageserver config as a feature flag for the new caching? We'll probably want to leave it in pre-prod for a bit to check perf impact and then release to prod when we're happy with it.

Reply:
I wonder if we need some special feature flag for it. If I replace `REL_SIZE_CACHE_CAPACITY` with parameter, then  we can set it to zero to disable pitr cache. Does it work?

Reply:
Well if we do that, we get a different behaviour from the current one.

Reply:
Oke - let's leave this as is. It will have a week to soak in staging.